StashiKu/aws-shop-be
# AWS-BE-Task-4-Integration-with-NoSQL-db

### link to FE deploy - https://d3ar6ba8glakkn.cloudfront.net/
### link to PR for FE part - https://github.com/StashiKu/shop-angular-cloudfront/pull/3

### [product-service ]  👈  (click to see open api docs)

  ### Functionality:
 **Reviewers should verify the lambda functions by invoking them through provided URLs.**
  - `getProductById` **[TransactionGetItems]** lambda returns product by id https://arq0plgxm6.execute-api.eu-west-1.amazonaws.com/dev/products/{id}
  - `getProductsList` **[Scan Request]** lambda returns a list of products https://arq0plgxm6.execute-api.eu-west-1.amazonaws.com/dev/products
  - `createProduct` **[TransactionPutItems]** lambda creates new product https://arq0plgxm6.execute-api.eu-west-1.amazonaws.com/dev/products
 
### How to test
 1. **Create Product**
  - try to hit endpoint with body like `{"description": "test product creation","count": 5, "price": 25, "title": "Product created by reviewer"}` - product should be created. Try to find this product on FE deploy
  -  then try to modify body, e.x remove one one of the parameters or change string to num - you should see specific error with status `400`
  -  try to hit endpoint without body - you should see specific error with status `400`
  2. **getProductsList** 
  - hit particular endpoint - you should see a list of products, copy id for the next step
  3. **getProductsById** 
  - hit particular endpoint - you should see requested product
  - hit endpoint without `id` param - you should get specific error
  - hit endpoint with `id` `e71116da-4c5d-11ed-bdc3-0242ac120002` to test case when record with such id exists in Prodcts table, but not in Stocks - you should see specific error and also this record should not appear in `getProductsList` request result
  4. click here https://github.com/StashiKu/shop-angular-cloudfront/pull/3 to continue with FE part
  
 ### More info:
 - `OpenApi` docs were generated with `serverless-auto-swagger` plugin. 
 - `Jest` is used for testing. Run `npm run test` to run tests
 - `Webpack`
 - `TypeScript`
 - `eslint`
 - `joi` validation
 
[product-service]: <https://cux94n1pu4.execute-api.eu-west-1.amazonaws.com/swagger>

**link to requirements** https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/4_integration_with_nosql_database/task.md

### Evaluation criteria
[**1**] - Task 4.1 is implemented
[**3**]- Task 4.2 is implemented lambda links are provided and returns data
[**4**]- Task 4.2 is implemented lambda links are provided and returns data
[**5**]- Your own Frontend application is integrated with Product Service (/products API) and products from Product Service are represented on Frontend. Link to a working Frontend application is provided for cross-check reviewer.

**Additional (optional) tasks**
[**+1**] - POST /products lambda functions returns error 400 status code if product data is invalid
[**+1**] - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
[**+1**] - All lambdas do console.log for each incoming requests and their arguments
[**+1**] - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa)
[**-**] - Rds

### Total score - 9

**Thanks for reviewing** 👍 

### ⚠️ PS: if approve, please write a comment below "Approved"
